src/compiler/checker.ts

```diff
@@ -20284,13 +20304,40 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (!result) {
                 const newMapper = createTypeMapper(root.outerTypeParameters, typeArguments);
                 const checkType = root.checkType;
-                const distributionType = root.isDistributive ? getReducedType(getMappedType(checkType, newMapper)) : undefined;
+                let distributionType = root.isDistributive ? getReducedType(getMappedType(checkType, newMapper)) : undefined;
+                let narrowingBaseType: Type | undefined;
+                const forNarrowing = distributionType && isNarrowingSubstitutionType(distributionType) && isNarrowableConditionalType(type, mapper);
+                if (forNarrowing) {
+                    narrowingBaseType = (distributionType as SubstitutionType).baseType;
+                    distributionType = getReducedType((distributionType as SubstitutionType).constraint);
+                }
                 // Distributive conditional types are distributed over union types. For example, when the
                 // distributive conditional type T extends U ? X : Y is instantiated with A | B for T, the
                 // result is (A extends U ? X : Y) | (B extends U ? X : Y).
-                result = distributionType && checkType !== distributionType && distributionType.flags & (TypeFlags.Union | TypeFlags.Never) ?
-                    mapTypeWithAlias(distributionType, t => getConditionalType(root, prependTypeMapping(checkType, t, newMapper), forConstraint), aliasSymbol, aliasTypeArguments) :
-                    getConditionalType(root, newMapper, forConstraint, aliasSymbol, aliasTypeArguments);
+                if (distributionType && checkType !== distributionType && distributionType.flags & (TypeFlags.Union | TypeFlags.Never)) {
+                    if (narrowingBaseType) {
+                        result = mapType(
+                            distributionType,
+                            (t: Type) =>
+                                getConditionalType(
+                                    root,
+                                    prependTypeMapping(checkType, getSubstitutionType(narrowingBaseType, t, /*isNarrowed*/ true), newMapper),

```

Comment:
I think the changes here aren't mentioned in your implementation notes. Why do we need custom logic for distribution when we're narrowing?

Reply:
Good question.

We are using substitution types (which can be thought of as `T & A`, where `T` is the base type and `A` the constraint) to convey a type parameter's narrowing. When instantiating a distributive conditional type in `getConditionalTypeInstantiation`, if the conditional's check type is a substitution type like `T & (A | B)`, the old logic would not distribute over this type, because it's a substitution type and not a union type. So, for distribution to happen, we have to take apart the `T & (A | B)` into `(T & A) | (T & B)`, and distribute over that.

The other special thing that we have to do is to take the intersection of the distribution result, as opposed to the union. This is because, if we narrow a type parameter `T` to `A | B`, and we have a conditional return type `T extends A ? R1 : T extends B ? R2 : T extends C ? R3 : never`, then we don't know which branch of the conditional return to pick, if branch `T extends A ? R1`, or branch `T extends B ? R2`, so we have to check whether the return expression's type is assignable to both, i.e. assignable to `R1 & R2`.

A semi-related note is that validating whether the conditional type has the right shape to be narrowed happens on-demand (and also at first as an optimization in `checkReturnExpression` when we're deciding whether to narrow the return type). This is because, as we instantiate a type, we may produce new conditional types that we need to then decide are safe to narrow or not (see [nested types case in `dependentReturnType6.ts` test](https://github.com/microsoft/TypeScript/pull/56941/files#diff-f40212beae0a9723380268c8a63a1560e91b9431a1fda539b2092cd4778e5fcfR28-R51)).

There's a bunch of details I haven't included in the implementation notes because I don't have a full list of them 😅. I'll add those.

Reply:
Would it make sense to always normalize union-containing substitution types into unions of substitution types before distributing?

Reply:
I haven't thought about it, but I assume it would break something somewhere in weird ways.

Reply:
it could help with some `NoInfer` problems: https://github.com/microsoft/TypeScript/pull/60271#discussion_r1806069774 :p

Reply:
I have this vague sense that it would be nice if the conditional-narrowing-specific logic didn’t affect the implementation of `getConditionalTypeInstantiation`, but even if it always normalized substitution types, I don’t have a suggestion for how to extract `mapTypeToIntersection`, so I guess it’s moot.